### DISCUSSION TITLE
Quarkus webauthn

### DISCUSSION THREAD

**Discussion by madocx** [quarkusio/quarkus#37746] [Q&A]
Working on a webauthn poc, utilizing webauthn4j (https://github.com/webauthn4j/webauthn4j). Everything works perfectly running in a JVM, but after compiling natively, the library fails verification of the signature. Specifically, com.webauthn4j.validator.AssertionSignatureValidator.verifySignature().

Under the covers, it seems webauthn4j is using JCASecurityServices. However, none of the recommendations here seem to help : https://www.graalvm.org/latest/reference-manual/native-image/dynamic-features/JCASecurityServices/

Has anyone else faced something like this, or have any recommendations on what might possibly be causing it? This is the first time I've run into something like this where it natively compiles without any issues, but just behaves entirely differently (and incorrectly) at runtime.

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
@madocx Can you consider experimenting with the Quarkus extension, see its https://quarkus.io/guides/security-webauthn description, it needs some updates, main one is that the tutorial code is now missing, but @FroMage put a lot of effort into it, it is based on Vert.x WebAuthn code, and it does compile in native. IMHO it would be nice to have this extension evolved a bit more.
Also note, prod quality OIDC providers will support WebAuthn.

**Reply by madocx** [reply]
So we definitely considered the quarkus extension, but had some hesitancy due to the preview status. Don't want to hitch our wagon to something that may or may not continue to be supported.

Our solution is using Cognito which doesn't offer any support for WebAuthn, other than the ability to implement it yourself via a custom auth flow.

**Reply by sberyozkin** [reply]
@madocx Sure, but with your own solution, support wise, it is your team which supports it completely :-). I support your choice, but would like to encourage you to play with the quarkus extension at some point, thanks

**Reply by madocx** [reply]
haha, fair point. If we can't find a quick solution to the webauthn4j issues we'll give it a go.

**Reply by madocx** [reply]
So we started kicking the tires on the quarkus extension, and have a crude implementation working both in jvm and natively - so thank you for the suggestion! However, we have a couple questions we were hoping to get some clarity on.

1. The below error message is showing up during startup. This key appears to be used to encrypt the response cookies, but documentation does not indicate how this should be configured. Is there a recommended way to configure the encryption key?

_2024-01-22 15:08:44,864 WARN  [io.qua.sec.web.WebAuthnRecorder] (Quarkus Main Thread) Encryption key was not specified for persistent WebAuthn auth, using temporary key MYgEYKXrSlQFEvIfwnf8uJspTJ7EWYPjP3ULBnp+YAU=_


2. Is it possible to have the cookie names used by the extension be configurable? Currently they're prefixed with "_quarkus_" which is somewhat bad practice for any public facing application, as it exposes technical details about our stack to end users.


3. We're implementing custom endpoints for challenge, register, and login. It looks as though there are nice convenience methods for register and login in https://github.com/quarkusio/quarkus/blob/main/extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnSecurity.java but not one for challenge. I imagine this is because there's not much need to customize this part of the process. This is true if the endpoints and cookies being used were configurable, but if there are any technical limitations on this, would there be any opposition to adding a challenge method to WebAuthnSecurity? Alternatively, we're having to implement this using the vertx classes directly.

**Reply by sberyozkin** [reply]
Thanks @madocx for deciding to give it a try, @FroMage, Steph,  can you please advise @madocx, see above.  Also CC-ing @michalvavrik since Michal updated the quickstart

**Reply by FroMage** [reply]
Hi.

1. You can configure the encryption key with `quarkus.http.auth.session.encryption-key` (I agree this should be documented)
2. You can configure the cookie name with `quarkus.webauthn.cookie-name` (this is in the docs at https://quarkus.io/guides/security-webauthn#quarkus-security-webauthn_quarkus.webauthn.cookie-name)
3. For the challenge, you mean the `/q/webauthn/callback` endpoint? Do you need to customise its code? Or are you looking for a way to change the predefined uris? Or both?

**Reply by FroMage** [reply]
Actually, there are two more cookies that are not customizable yet. I'll open an issue.

**Reply by madocx** [reply]
Excellent, thanks FroMage! 

Yeah sorry for the confusion, I should have specified it was the other cookies we were concerned with.

Regarding callback endpoint, customization of the uri would be nice but we'll probably end up implementing our own so that we can short-circuit the process for business rules given its the first step. I assume we'd want to invoke all the webauthn steps similarly through WebAuthnSecurity.java but will defer to you if you think there's a better way.

**Reply by FroMage** [reply]
I'm sorry, I don't understand exactly what you're trying to do :(

**Reply by madocx** [reply]
No worries, that's my fault. After reviewing the documentation again realized I was misunderstanding the order of operations on the endpoints. Since /q/challenge isn't invoked until after /q/login or /q/register, we really don't have any need for a custom implementation as long as the URI is configurable.

**Reply by FroMage** [reply]
Yeah, the names are a bit confusing.

- `/q/webauthn/login` gets you a login challenge
- `/q/webauthn/register` gets you a registration challenge
- `/q/webauthn/callback` authenticates you, with a login or registration body. You can override this endpoint and implement it yourself with the help of `WebAuthnSecurity`

At the moment, you can't easily override the challenge endpoints, because they reimplement some of the Vert.x WebAuthn logic to bypass the lack of sessions (using cookies).

If you intend to override the `/q/webauthn/callback` endpoint (like we document you can do), then you don't need to customise any URI, you can place your endpoint where you want, since you're the ones calling them, Quarkus WebAuthn doesn't have to know where they are.

If you intend to override the `/q/webauthn/register|login` challenges, then you'll probably need that I expose their equivalents in `WebAuthnSecurity` to deal with the session cookies.

If you don't, and you just want them to be registered on different URIs, then that's a configuration option we need to add.

So, is this clearer? What do you want, then? :)

**Reply by madocx** [reply]
I was with you half way through, but then I got lost - probably because I don't understand the inner workings of the extension. regiseter/login are already exposed in WebAuthnSecurity from what I see in the source. Am I missing something about what you're suggesting? https://github.com/quarkusio/quarkus/blob/main/extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnSecurity.java

**Reply by FroMage** [reply]
Yeah, the names are confusing, I should have changed all that at the start.

The `/q/webauthn/login` endpoint gets you a _challenge_ that you need on the client in order to call an authenticator, and then you trigger the login _action_ by calling the `/q/webauthn/login` endpoint.

So, one is the challenge, followed by the action. The two challenges have separate endpoints, while the two actions share the same endpoint. By default.

`WebAuthnSecurity.login` is an override of the _action_, not the challenge. In case you want to have different actions, or customise what you do in the action. ATM there's no way to let you write custom challenges.
